GTG/backends/bugzilla/bugzilla.py

```diff
@@ -0,0 +1,201 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2009 - Guillaume Desmottes <gdesmott@gnome.org>
+#
+# This program is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+import threading
+import socket
+
+from xmlrpc.client import Fault as XmlrpcFault
+
+from collections import namedtuple
+from urllib.parse import urlparse
+
+from gi.repository import GObject
+
+from GTG.core.translations import _
+from GTG.tools.logger import Log
+
+from GTG.backends.bugzilla import exceptions
+from GTG.backends.bugzilla.notification import send_notification
+from GTG.backends.bugzilla.services import create_bugzilla_service
+
+__all__ = ('BugInformationSyncTask',)
+
+BugSyncTaskInfo = namedtuple('BugSyncTaskInfo',
+                             ['scheme', 'net_location', 'bug_id'])
+BUG_URL_PATTERN = re.compile(r'^(https?)://(.+)/show_bug\.cgi\?id=(\d+)$')
+
+BUGZILLA_NO_PERM_MESSAGE = _("You have no permission to read bug <b>%s</b>.")
+BUGZILLA_INVALID_MESSAGE = _("Bug ID <b>%s</b> is invalid")
+BUGZILLA_NOT_EXIST_MESSAGE = _("Bug ID <b>%s</b> does not exist.")
+BUGZILLA_BUG_SYNC_FAIL_MESSAGE = _("Failed to synchronize information for "
+                                   "bug <b>%s</b> due to some unknown reason.")
+
+
+def parse_bug_url(url):
+    '''
+    Extract URL data
+
+    @param url: a string representing a URL
+    @return: a tuple containing scheme, hostname, and a list of key-value
+             queries
+    '''
+    r = urlparse(url)
+    queries = dict([item.split('=') for item in r.query.split('&')])
+    return r.scheme, r.netloc, queries
+
+
+def convert_to_list(value):
+    if isinstance(value, (list, tuple)):
+        return list(value)
+    else:
+        return [value]
+
+
+class BugInformationSyncTask(threading.Thread):
+
+    def __init__(self, task, backend, **kwargs):
+        ''' Initialize task data, where task is the GTG task object. '''
+        self.task = task
+        self.backend = backend
+        super().__init__(**kwargs)
+
+    def _collect_tags(self, bug):
+        tags = []
+        parameters = self.backend.get_parameters()
+
+        if parameters['bugzilla-tag-use-priority']:
+            tags += convert_to_list(bug.priority)
```

Comment:
Different Bugzilla service may return different attribution of a bug. For example, component of a bug returned from `bugzilla.redhat.com` is a list containing at least one component name, whereas it is a string value returned from `bugzilla.gnome.org`. Whatever a type of each attribution depends on the concrete Bugzilla service, and it is unpredictable for GTG or any other users.

`convert_to_list` is a defensive way to overcome this situation to the attributions used in Bugzilla synchronization service.